### PR TITLE
Prepared statement cache retrofit

### PR DESCRIPTION
**Pull request by BillyYccc**
1. move prepared statement cache from `pgclient` to `sqlclient` so that we can use it in `mysqlclient`
2. introduce a `LruCache` for the cache implementation
3. add support for configuring with cache size and cache sql length limit
4. close the statement when it's evicted from the cache

### CODE REVIEW

**Comment by BillyYccc** [comment]
Using cache will somehow alleviate producing `Can't create more than max_prepared_stmt_count ` error stated in #314 when preparing a lot of statements with the same sql string.

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/SqlConnectOptions.java [hunk 1]
```diff
@@ -9,11 +9,18 @@
...
+  public static final int DEFAULT_PREPARED_STATEMENT_CACHE_SQL_LIMIT = 2048;
+
   private String host;
   private int port;
   private String user;
   private String password;
   private String database;
+  private boolean cachePreparedStatements = DEFAULT_CACHE_PREPARED_STATEMENTS;
+  private int preparedStatementCacheSize = DEFAULT_PREPARED_STATEMENT_CACHE_SIZE;
+  private int preparedStatementCacheSqlLimit = DEFAULT_PREPARED_STATEMENT_CACHE_SQL_LIMIT;
```
Comment (vietj) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/SqlConnectOptions.java]: rename preparedStatementCacheMaxSize

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/LruCache.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package io.vertx.sqlclient.impl;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * A reusable cache implementation based on LRU(Least Recently Used) strategy.
+ */
+public class LruCache<K, V> extends LinkedHashMap<K, V> {
```
Comment (vietj) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/LruCache.java]: make this package private

Comment (vietj) [reply vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/LruCache.java]: I think it's actually not need to have this intermediary class, just make the cache class extends `LinkedHashMap`

**Pull request merged**